<launch>
    <!-- Arguments -->
    <arg name="ns_prefix" default="uav_"/>
    <arg name="robot_model" default="iris"/>
    <arg name="mode" default="sitl"/>
    <arg name="multi" default="false"/>
    <arg name="pub_rate" default="100.0"/>
    <arg name="path" default="cubic"/>
    <arg name="reach_tolerance" default="0.1"/>

    <node pkg="rviz" type="rviz" name="rviz_node" args="-d $(find upat_follower)/rviz/mision.rviz" required="true"/>

    <!-- Launch Gazebo simulation -->
    <rosparam param="/use_sim_time">true</rosparam>
    <node pkg="px4_bringup" type="launch_gzworld.py" name="gazebo_world" output="screen" args="-physics=ode -world=$(find upat_follower)/tests/experiments/worlds/karting_sevilla.world -add_model_path=$(find upat_follower)/tests/experiments/models -description_package=robots_description">
        <rosparam param="sim_origin">[37.558542, -5.931074, 7.89]</rosparam>        <!-- [lat,lon,alt] -->
    </node>

    <!-- Launch robot id=1 -->
    <include file="$(find px4_bringup)/launch/spawn_robot.launch" ns="$(arg ns_prefix)1">
        <arg name="robot_model" value="$(arg robot_model)"/>
        <arg name="id" value="1"/>
        <arg name="material" value="Yellow"/>
        <arg name="mode" value="$(arg mode)"/>
        <arg name="x" value="-32.11"/>
        <arg name="y" value="4.01"/>
        <arg name="z" value="0.0"/>
        <arg name="yaw" value="1.57"/>
    </include>

    <!-- Launch UAL server for drone id=1 -->
    <node pkg="uav_abstraction_layer" type="server" name="ual" output="screen" ns="$(arg ns_prefix)1">
        <param name="uav_id" value="1"/>
        <rosparam param="home_pose">[-32.11, 4.01, 0]</rosparam>        <!-- [x,y,z] -->
    </node>

    <!-- UAV Path Manager nodes -->
    <group ns="upat_follower">
        <group ns="$(arg ns_prefix)1">
            <node pkg="upat_follower" type="ual_communication_node" name="ual_communication" output="screen" required="true">
                <param name="uav_id" value="1"/>
                <param name="save_test_data" value="false"/>
                <param name="pub_rate" value="$(arg pub_rate)"/>
                <param name="path" value="$(arg path)"/>
                <param name="reach_tolerance" value="$(arg reach_tolerance)"/>
                <param name="use_class" value="true"/>
            </node>
            <node pkg="upat_follower" type="visualization_node" name="visualization" required="true" output="screen">
                <param name="uav_id" value="1"/>
                <param name="robot_model" value="$(arg robot_model)"/>
                <param name="pub_rate" value="$(arg pub_rate)"/>
            </node>
        </group>
    </group>
</launch>
